File: src/main/java/org/lovaprod/model/Wallet.java
package org.lovaprod.model;

import org.lovaprod.utils.MoneyUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class Wallet {
    private String id;
    private double balance;
    public static double MAX_TRANSACTION_AMOUNT=200_000_000;
    private User account;
    private TransactionHistory transactionHistory;

    public Wallet() {
        this.id = UUID.randomUUID().toString();
        this.balance = 0.0;
        this.transactionHistory = new TransactionHistory();
    }

    public String getId() {
        return id;
    }

    public double getBalance() {
        return balance;
    }


    public User getAccount() {
        return account;
    }

    public void setAccount(User account) {
        this.account = account;
    }

    public TransactionHistory getTransactionHistory() {
        return transactionHistory;
    }

    public void deposit(double amount) {
        if (MoneyUtils.isAmountValid(amount)){
            balance += amount;
            transactionHistory.addTransaction(amount, Transaction.TransactionType.DEPOSIT, "me");
        }
    }

    public boolean withdraw(double amount) {
        if(MoneyUtils.isAmountValid(amount, balance, MAX_TRANSACTION_AMOUNT)){
            balance -= amount;
            transactionHistory.addTransaction(amount, Transaction.TransactionType.WITHDRAW, "");
            return true;
        }
        return false;
    }

    public boolean withdraw(double amount, User destinationUser) {
        if(MoneyUtils.isAmountValid(amount, balance, MAX_TRANSACTION_AMOUNT)){
            balance -= amount;
            destinationUser.getWallet().deposit(amount);
            transactionHistory.addTransaction(amount, Transaction.TransactionType.TRANSFER, destinationUser.getUsername());
            return true;
        }
        return false;
    }

}

File: Wallet.cs
using System;
using System.Linq;

namespace WalletApp
{
    public class Wallet
    {
        public string Id { get; }
        public double Balance { get; private set; }
        public static double MaxTransactionAmount = 200_000_000;
        public User Account { get; set; }
        public TransactionHistory TransactionHistory { get; }

        public Wallet()
        {
            Id = Guid.NewGuid().ToString();
            Balance = 0.0;
            TransactionHistory = new TransactionHistory();
        }

        public void Deposit(double amount)
        {
            if (MoneyUtils.IsAmountValid(amount))
            {
                Balance += amount;
                TransactionHistory.AddTransaction(amount, Transaction.TransactionType.Deposit, "me");
            }
        }

        public bool Withdraw(double amount)
        {
            if (MoneyUtils.IsAmountValid(amount, Balance, MaxTransactionAmount))
            {
                Balance -= amount;
                TransactionHistory.AddTransaction(amount, Transaction.TransactionType.Withdraw, "");
                return true;
            }
            return false;
        }

        public bool Withdraw(double amount, User destinationUser)
        {
            if (MoneyUtils.IsAmountValid(amount, Balance, MaxTransactionAmount))
            {
                Balance -= amount;
                destinationUser.Wallet.Deposit(amount);
                TransactionHistory.AddTransaction(amount, Transaction.TransactionType.Transfer, destinationUser.Username);
                return true;
            }
            return false;
        }
    }
}
